Migrate to `Compose`

alesrazym opened the pull request:
Starting migration to Jetpack Compose, as per https://developer.android.com/jetpack/compose/migrate

Commit:
WIP: Update Kotlin, AGP, Compose

Update:
AGP to 8.1.3
Kotlin to 1.9.20
Compose BOM to 2023.10.01
Androidx.core to 1.12.0

Commit:
WIP: Replace rockets list and adapter with Compose

Commit:
WIP: Continue replace rockets list and adapter with Compose

Rfc list item to compose `Row` and `Column` instead of view's `ConstraintLayout`
Rfc recycler view to `LazyColumn`

Commit:
REF: Rfc rocket list to `composables`

Add `androidx.compose.material` to have pull to refresh.
Rfc `SwipeRefreshLayout` to composables.
Rfc `ScreenUiState` to not contain interface `UiState`, but individual properties.
Rfc view model to not initialize with `init`, but later with `initialize()`
Rfc view model state collecting from `flowWithLifecycle` to `by .collectAsState` in Composable.
Update `RocketListViewModelTest`.

Commit:
REF: Rfc rest of rocket list to `composables`

Remove `swiperefreshlayout` dependency.
Remove `fragment_rocket_list.xml`.
Remove `FragmentRocketListBinding` from fragment, replace by one compose view.
Rfc composables, add rocket list title.

alesrazym commented:
rebased on develop (with detail merged) and fix

Commit:
REF: Rename RocketList composable to `RocketListFragmentContent`

Commit:
WIP: Rfc RocketDetailFragment's first part to compose

Commit:
LIB: Update `activity-compose:1.8.1`

Commit:
REF: Rfc parameter card to compose

Commit:
REF: Rfc rest of detail view to compose, except correct view model handling and toolbar.

Commit:
REF: Rfc detail view's toolbar to compose.

Remove `fragment_rocket_detail`, `rocket_detail_menu` xmls and binding.
Add argument to navigation to temporarily keep rocket name.

Commit:
REF: Rfc rocket launch fragment (placeholder) to compose

Commit:
REF: Rfc fragments to Compose screens, navigation and DI

Rfc fragments and activity to compose only.
Rfc navigation graph to compose, create Screen class.

Remove unused and add new compose dependencies

Commit:
REF: Rfc Screen and navigation graph to individual Screens

Create `NavGraphBuilder.xxxScreen` extension methods, hoist state (and view model) there.
Create `NavController.navigateToRocketDetail` extension for nav args safety

Inspired by https://developer.android.com/guide/navigation/design/type-safety and some of sample repos

Commit:
REF: Change `getFirstFlightFormat` property to getter function to avoid locale change problems

Commit:
WIP: Theme & style

Commit:
WIP: Theme & style detail screen, code style

Commit:
FIX: Fix preview rocket id in list

Commit:
REF: Theme & style detail screen, rfc colors from `MaterialTheme` to `staticCompositionLocalOf`

Commit:
REF: Lazy init colors

havlisimo commented on app/src/main/java/cz/quanti/razym/rocketapp/presentation/RocketDetailViewModel.kt:
```diff
@@ -11,40 +11,50 @@ import cz.quanti.razym.rocketapp.model.RocketDetail
+                            it.copy(
+                                loading = false,
+                                messages = it.messages.apply {
+                                    add(result.exception?.message ?: "Unknown error")
```
1. pouzivani message z exception pro ui state je spatne, vetsinou tam nebyva nic co by bylo nejak uzivatelsky citelne, nehlede na to ze to neni lokalizovane
2. unknown error by melo byt ve strings
3. vsiml jsem si ze se ta hodnota na UI stejne nepouziva, ale nejaky error handling by to chtelo

havlisimo commented on app/src/main/java/cz/quanti/razym/rocketapp/system/RocketDetailScreen.kt:
```diff
@@ -0,0 +1,541 @@
+            TopBar(
+                // TODO, because we does not persist data, use a placeholder for title
+                //  when not loaded yet.
+                title = rocket?.name ?: "Loading",
```
use string resource for Loading or send the title as a navigation parameter

alesrazym replied:
used both

havlisimo commented on app/src/main/java/cz/quanti/razym/rocketapp/system/RocketDetailScreen.kt:
```diff
@@ -0,0 +1,541 @@
+        Box(
+            modifier = Modifier
+                .fillMaxSize()
+                .background(LocalColors.current.primaryContainer)
```
chtel bych abys udelal object AppTheme, kde bude AppTheme.color a tak se k tomu bude pristupovat, LocalColors bude private

havlisimo commented on app/src/main/java/cz/quanti/razym/rocketapp/ui/theme/Theme.kt:
```diff
@@ -1,70 +1,80 @@
 package cz.quanti.razym.rocketapp.ui.theme
```
prosim vytvorit modul ui, kam se daji vsechny veci tykajici se nastaveni theme a sdilene composable komponenty (napr. ContentStatusText)

havlisimo commented on app/src/main/java/cz/quanti/razym/rocketapp/system/RocketDetailScreen.kt:
```diff
@@ -0,0 +1,541 @@
+
+@Composable
+fun formatReusable(reusable: Boolean): String {
+    return if (reusable) {
```
UI state by mel rovnou obsahovat string resource id, abychom se vyhli podmince v UI

havlisimo commented on app/src/main/java/cz/quanti/razym/rocketapp/system/RocketDetailScreen.kt:
```diff
@@ -0,0 +1,541 @@
+
+@Composable
+fun formatBurn(burnSec: Int?): String {
+    return if (burnSec == null) {
```
UI state by mel rovnou obsahovat string resource id, abychom se vyhli podmince v UI

alesrazym replied:
Vytvoril jsem UiText wrapper. Pokud je to dobre reseni, prevedu i dalsi casti kodu

havlisimo commented on app/src/main/java/cz/quanti/razym/rocketapp/system/RocketDetailScreen.kt:
```diff
@@ -0,0 +1,541 @@
+        modifier = Modifier
+            .fillMaxWidth(),
+    ) {
+        flickrImages.take(10).forEach {
```
state z view modelu by uz mel obsahovat pripraveny list, aby se tady nemuselo delat take(10), to by mel prave udelat VM

alesrazym replied:
vyhozeno z UI, ale TODO kde to umistit

havlisimo commented on app/src/main/java/cz/quanti/razym/rocketapp/system/RocketDetailScreen.kt:
```diff
@@ -0,0 +1,541 @@
+        Icon(
+            modifier = Modifier
+                .size(40.dp)
+                .padding(8.dp)
```
prosim vytvorit AppTheme.dimens, kam se schovaji vsechny hodnoty pro dp

havlisimo commented on app/src/main/java/cz/quanti/razym/rocketapp/system/RocketDetailScreen.kt:
```diff
@@ -0,0 +1,541 @@
+                } else {
+                    RocketDetailContent(rocket)
+                }
+            }
```
doporucuju naimplementovat nejakou sdilenou composable, ktera bude handlovat loading a error state obecne pro vsechny screeny
napr
```
@Composable
fun StateFullContent(
    state: ScreenState, 
    modifier: Modifier, 
    onErrorClick: () -> Unit, 
    content: () -> Unit
) {
    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(LocalColors.current.primaryContainer)
            .padding(innerPadding)
    ) {
        when (state) {
            ScreenState.LOADING -> {
                ContentStatusText(
                    text = R.string.getting_rocket_detail_in_progress,
                )
            }
            ScreenState.ERROR -> {
                ContentStatusText(
                    text = R.string.getting_rocket_detail_failed,
                    onClick = onErrorClick,
                )
            }
            ScreenState.CONTENT -> {
                content()
            }
        }
    }
}
```

Commit:
REF: Let UI state to handle string resources.

Create `UiText` wrapper around different string resources, so that logic is in view model, but view model does not need context.
Create additional data classes in `presentation` package to provide `UiText`.
Adjust test and detail screen.
TODO comments left for issues to be discussed.

Commit:
REF: Move limit of photos from UI.

TODO comments left for issues to be discussed.

Commit:
REF: WIP custom `RocketappTheme`.

Add RocketappTheme object.
Add dimensions to theme, rfc usage.
Make LocalColors private, rfc usage.

Commit:
FIX: Fix smaller images to be `Crop`ped

Commit:
REF: Extract "Loading" to resources

Commit:
REF: Add rocket name to navigation optional argument.

Commit:
REF: Use `collectAsStateWithLifecycle` instead of `collectAsState` on ViewModels' Flows

Commit:
REF: Add `Modifier` to `ContentStatusText` param, move to `ui` package

Commit:
REF: Add `RocketAppTypography`, remove `MaterialTheme`.

Fix `RocketLaunchScreen` (rename from fragment, add dimens)

Commit:
FIX: Fix some TODOs, code cleanup

Commit:
REF: Use extension fun instead of secondary ctor

Commit:
REF: Add `Shapes` to `RocketappTheme`

Commit:
Fix: Use `RocketappTheme` in `ContentStatusText`

Commit:
Fix: Replace `material.Text` with `material3.Text`

Commit:
Code style

Commit:
REF: Make `ContentStatusText` to accept `UiText` instead of @StringRes. Add `UiText.Empty` for convenience (instead of null)

Commit:
REF: Create `UiScreenState` to be a generic state holder for pull to refresh views.

Add `PullToRefresh` reusable composables.
Add Exception handling in `RocketListViewModel`, add string resources.
Rfc `RocketListViewModelTest` to match changes, add `Rocket` equity functions.
Rfc `RocketListScreen` to match changes.

Commit:
REF: Rfc `RocketDetailScreen` to use `UiScreenState`.

Commit:
WIP: Add exception handling to Screens using simple `SnackBar`

Commit:
FEAT: Add custom `@RocketAppPreview` and use `@PreviewParameter`. Update previews

Commit:
REF: Rfc `RocketAppTypography`.

`RocketAppTypography` to use `TextStyle` directly instead of `copy()`.
Adjust not to use `copy()` in Composable.
Also fix some sizing and weights to be as per design.

Commit:
REF: Rename `UiScreenState.Success` to `UiScreenState.Data`

Commit:
REF: Make default errorTransform in `UiScreenState<*>.update` more robust, remove from view models

Commit:
FEAT: Add data param to `successContent` lambda of `StateFullPullToRefresh`

Commit:
FIX: Fix `defaultErrorTransform` to work as expected, fix `RocketDetailViewModelTest` tests

Commit:
FEAT: Add error message UI with `SnackBar` using `CompositionLocal`

Commit:
Cleanup

Commit:
REF: Move common UI code to `ui` module

Commit:
REF: Split `Theme` into Colors.kt, Dimens.kt, Shapes.kt and Typography.kt

alesrazym commented:
@havlisimo Doufam, ze jsem zapracoval vsechny komenty a to, na cem jsme se domluvili. Rebasnul jsem vetev na develop a pripravil druhy PR https://github.com/alesrazym/rocket-app/pull/12. Puvodni vetev jsem nechal zatim pro kontrolu historie, pripadne kdybychom se chteli k necemu vratit. Tenhle PR zavru az zavreme ten novy
